libcextract/SymbolExternalizer.cpp

```diff
@@ -71,142 +87,161 @@ using namespace llvm;
-    TypeUpdaterVisitor(SymbolExternalizer &se, ValueDecl *new_decl,
-                       const std::string &old_decl_name, bool wrap)
-    : SE(se),
-      NewSymbolDecl(new_decl),
-      OldSymbolName(old_decl_name),
-      Wrap(wrap)
+    TypeUpdaterVisitor(SymbolExternalizer &se)
+    : SE(se)
     {}
 
   enum {
     VISITOR_CONTINUE = true,   // Return this for the AST transversal to continue;
     VISITOR_STOP     = false,  // Return this for the AST tranversal to stop completely;
   };
 
-  /* The updator method.  This will be called by the Visitor when traversing the
-     code.  */
-  bool VisitTypeOfExprType(TypeOfExprType *type)
+  bool VisitDeclaratorDecl(DeclaratorDecl *decl)
   {
-    /* Create a instance of our tested-in-battle FunctionUpdater...  */
-    SymbolExternalizer::FunctionUpdater fu(SE, NewSymbolDecl, OldSymbolName, Wrap);
+    SymbolUpdateStatus *sym = SE.getSymbolsUpdateStatus(decl->getName());
+    if (sym == nullptr)
+      return VISITOR_CONTINUE;
+
+    ExternalizationType type = sym->ExtType;
+    if (type == ExternalizationType::STRONG) {
+      if (!sym->Done) {
+        std::string sym_name = decl->getName().str();
+
+        /* If we found the first instance of the function we want to externalize,
+           then proceed to create and replace the function declaration node with
+           a variable declaration node of proper type.  */
+        std::string new_name = EXTERNALIZED_PREFIX + sym_name;
+        sym->new_name = new_name;
+        DeclaratorDecl *new_decl = SE.Create_Externalized_Var(decl, new_name);
+        SE.Log.push_back({.OldName = sym_name,
+                       .NewName = new_name,
+                       .Type = type});
+
+        /* Create a string with the new variable type and name.  */
+        std::string o;
+        llvm::raw_string_ostream outstr(o);
+
+        /*
+         * It won't be a problem to add the code below multiple times, since
+         * clang-extract will remove ifndefs for already defined macros
+         */
+        if (SE.Ibt) {
+          outstr << "#ifndef KLP_RELOC_SYMBOL_POS\n"
+                    "# define KLP_RELOC_SYMBOL_POS(LP_OBJ_NAME, SYM_OBJ_NAME, SYM_NAME, SYM_POS) \\\n"
+                    "   asm(\"\\\".klp.sym.rela.\" #LP_OBJ_NAME \".\" #SYM_OBJ_NAME \".\" #SYM_NAME \",\" #SYM_POS \"\\\"\")\n"
+                    "# define KLP_RELOC_SYMBOL(LP_OBJ_NAME, SYM_OBJ_NAME, SYM_NAME) \\\n"
+                    "   KLP_RELOC_SYMBOL_POS(LP_OBJ_NAME, SYM_OBJ_NAME, SYM_NAME, 0)\n"
+                    "#endif\n\n";
+        }
 
-    /* ... and call the method which updates the body of a function. (but that
-       is not a function!  But who cares, and Expr is a special type of Stmt
-       in clang so everything works!  */
-    fu.Update_References_To_Symbol(type->getUnderlyingExpr());
+        new_decl->print(outstr);
 
-    return VISITOR_CONTINUE;
-  }
+        if (SE.Ibt) {
+          std::string sym_mod = SE.IA.Get_Symbol_Module(sym_name);
```

Comment:
`const std::string &sym_mod`

Reply:
This one can't be const since we can assign it later to vmlinux if it's empty.

Reply:
On both cases there is no need to store a copy of the string. Why can't it be const std::string &?